### PR TITLE
Fix WinHttpHandler error handling

### LINKED ISSUE #28156: System.NullReferenceException in System.Net.Http.WinHttpRequestCallback.RequestCallback
Couldn't find any issue that seemed to cover the problem. I'm running kestrel in tests, and writing a reverse proxy component.

At an unknown point in running the tests, I get the following exception:
```
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Http.WinHttpRequestCallback.RequestCallback(IntPtr handle, WinHttpRequestState state, UInt32 internetStatus, IntPtr statusInformation, UInt32 statusInformationLength)
   at System.Net.Http.WinHttpRequestCallback.WinHttpCallback(IntPtr handle, IntPtr context, UInt32 internetStatus, IntPtr statusInformation, UInt32 statusInformationLength)
   at Interop.WinHttp.WinHttpCloseHandle(IntPtr handle)
   at Interop.WinHttp.SafeWinHttpHandle.ReleaseHandle()
   at System.Runtime.InteropServices.SafeHandle.InternalDispose()
   at System.Net.Http.WinHttpRequestState.ClearSendRequestState()
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

Repro:
 - code: openrasta/openrasta-core@e2b2c5d
 - tests: https://github.com/openrasta/openrasta-core/blob/1e006fddfd91b14d92177c3948517a809074f553/src/Tests/Plugins.ReverseProxy/errors/proxied_timeout.cs
 - build failing: https://ci.appveyor.com/project/OpenRasta/openrasta-core/build/2.6.0-preview.1.1298+master

It looks like the same stacktrace as in #26278 

Note builds and tests succeed on core 2 on libcurl-based http client.

I tried not disposing request/response messages from the client, no change. I'm refactoring the code at the moment, so maybe the problem will go away, but i'm pretty confused.

### PR DESCRIPTION
**Pull request by davidsh**
This PR addresses some error handling problems in WinHttpHandler as well as
improves the error messages generated for WinHttpException.

This was root caused to a bug in WinHttp where it can't do a GET request with a
request body using chunked encoding. However, this exception was masked due to
an error handling bug caused by an inconsistency in how WinHttp associates
context values for callbacks during the WinHttpSendRequest API call. This PR now
fixes the error handling around synchronous errors being returned from
WinHttpSendRequest. The root WinHttp bug itself can't be fixed. So, doing a GET request
will still throw an exception, but it will be a catchable HttpRequestException.

Another bug was discovered while investigating #26278. WinHttpCloseHandle should only be
called once and should never race between threads. This is normally protected when
the request handle is closed via SafeHandle Dispose(). But there was a place in the
callback method where we called WinHttpCloseHandle directly against the raw handle.

Finally, the error messages for WinHttpExceptions have been improved to show the
error number and the probable WinHttp API call that generated the error. This will
save diagnostic time.

Example:

Original WinHttpException message: "The parameter is incorrect"
New WinHttpException message: "Error 87 calling WinHttpSendRequest, 'The parameter is incorrect'."

I also moved the WinHttpException source code back from Common. It was originally shared between WinHttpHandler and ClientWebSocket. But then ClientWebSocket moved away from WinHttp implementation. So, it makes more sense for this class to be under WinHttpHandler.

Closes #28156
Contributes to #26278

### CODE REVIEW

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs [hunk 1]
```diff
@@ -296,7 +296,7 @@ public void SaveServerCredentialsToCache(Uri uri, uint authScheme, ICredentials
                 Interop.WinHttp.WINHTTP_OPTION_AUTOLOGON_POLICY,
                 ref optionData))
             {
-                WinHttpException.ThrowExceptionUsingLastError();
+                WinHttpException.ThrowExceptionUsingLastError("WinHttpSetOption");
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs]: nit: `nameof(Interop.WinHttp.WinHttpSetOption)`... same in other such situations.  It'll result in the same string, just safe against refactorings and the like.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 2]
```diff
@@ -1343,11 +1343,13 @@ private RendezvousAwaitable<int> InternalSendRequestAsync(WinHttpRequestState st
                     0,
                     state.ToIntPtr()))
                 {
-                    // Dispose (which will unpin) the state object. Since this failed, WinHTTP won't associate
-                    // our context value (state object) to the request handle. And thus we won't get HANDLE_CLOSING
-                    // notifications which would normally cause the state object to be unpinned and disposed.
+                    // WinHTTP doesn't always associate our context value (state object) to the request handle.
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Is that a bug in WinHTTP?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs (hunk 1 above)
Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs]: Will fix.

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build

**Comment by davidsh** [comment]
Unrelated CI failures in System.ServiceModel.Syndication.Tests.BasicScenarioTests

**Pull request merged**